Various Cantherm UX improvements

alongd opened the pull request:
Various improvements to the user experience in Cantherm:
- Automatically read the spinMultiplicity from the QM job
- Automatically assign the frequency scale factor for the given model chemistry
- Simply use `Log` instead of `GaussLog` / `MolproLog` / `QChemLog`, while keeping backwards compatibility
- Allow adjLists when specifying structures (very helpful for heretoatoms with problematic SMILES)
- Updates to the documentation
- Updates to Cantherm's examples
- Additional exceptions with helpful debug messages
- Plot rates and thermo by default

cgrambow commented on rmgpy/cantherm/molpro.py:
```diff
@@ -28,9 +28,7 @@
-import math
-import numpy
-import os.path
+import math, numpy, logging
```
Keep imports on separate lines.

cgrambow commented on rmgpy/cantherm/pdep.py:
```diff
@@ -35,17 +35,12 @@
-import logging
-
-import rmgpy.constants as constants
+import math, numpy, logging
```
Keep imports on separate lines.

cgrambow commented on rmgpy/cantherm/pdep.py:
```diff
@@ -511,25 +506,30 @@ def plot(self, outputDirectory):
-                for p in range(Pcount):
-                    pylab.semilogy(1000.0 / Tlist, K[:,p], color=cm(1.*p/(Pcount-1)), marker='o', linestyle='')
+                for p in xrange(Pcount):
+                    pylab.semilogy(1000.0 / Tlist, K[:,p], color=cm(1.*p/(Pcount-1)), marker='o', linestyle='',
```
See previous comment about `pylab`.

cgrambow commented on rmgpy/cantherm/qchem.py:
```diff
@@ -28,11 +28,9 @@
 
-import math
-import numpy
+import math, numpy, logging
```
Keep imports on separate lines.

cgrambow commented on rmgpy/cantherm/statmech.py:
```diff
@@ -772,6 +792,41 @@ def applyEnergyCorrections(E0, modelChemistry, atoms, bonds,
+                f.close()
+                software = QchemLog(fullpath)
+                break
+            elif any(x in line for x in ['ATOMIC COORDINATES','PROGRAM']):
```
At least for Molpro 2015, it prints `MOLPRO` towards the top of the file. Should we check for that instead?

cgrambow commented on rmgpy/cantherm/statmech.py:
```diff
@@ -993,3 +1048,59 @@ def projectRotors(conformer, F, rotors, linear, TS):
+def assign_frequency_scale_factor(model_chemistry):
+    """
+    Assign the frequency scaling factor according to the model chemistry.
+    Sources: http://cccbdb.nist.gov/vibscalejust.asp, J. Chem. Phys. 1999, 110, 2822–2827
```
One thing to think about going forward is which values we want to assign as default. For example, [this website](https://comp.chem.umn.edu/freqscale/index.html) seems to be more actively maintained, but it's unclear which values are more reliable. The website has an explicit value for CCSD(T)-F12a/cc-pVTZ-F12, but not for other basis sets.

cgrambow commented on rmgpy/cantherm/thermo.py:
```diff
@@ -34,9 +34,8 @@
-import math
 import numpy.linalg
-import logging
+import logging, string
```
Keep imports on separate lines.

cgrambow commented on rmgpy/cantherm/thermo.py:
```diff
@@ -225,12 +224,12 @@ def plot(self, outputDirectory):
         pylab.xlabel('Temperature (K)')
         pylab.ylabel('Heat capacity (cal/mol*K)')
-        pylab.legend(['statmech', 'thermo'], loc=4)
+        pylab.legend(['statmech', 'fitted'], loc=4)
```
`pylab` comment

cgrambow commented on rmgpy/cantherm/thermo.py:
```diff
@@ -248,5 +247,10 @@ def plot(self, outputDirectory):
         fig.subplots_adjust(left=0.10, bottom=0.08, right=0.95, top=0.95, wspace=0.35, hspace=0.20)
-        pylab.savefig(os.path.join(outputDirectory, 'thermo.pdf'))
+
+        if not os.path.exists('plots'):
```
Make `plots` in `outputDirectory`?

alongd replied:
'outputDirectory' becomes too packed if one has many plots, so I think it's better to have it in a subfolder

alongd commented on rmgpy/cantherm/statmech.py:
```diff
@@ -249,7 +249,7 @@ def load(self):
         try:
             spinMultiplicity = local_context['spinMultiplicity']
         except KeyError:
-            raise InputError('Required attribute "spinMultiplicity" not found in species file {0!r}.'.format(path))
```
I'm not raising this error anymore, now the user is allowed not to input `spinMultiplicity`.
A different error is raised if the user did not supply a value *and* we couldn't find the multiplicity from the log file

alongd commented on rmgpy/cantherm/statmech.py:
```diff
@@ -778,6 +792,41 @@ def applyEnergyCorrections(E0, modelChemistry, atoms, bonds,
+    def __init__(self, path):
+        self.path = path
+
+    def determine_qm_software(self, fullpath):
```
I like the suggestion, but I think it's not straight forward to implement. Since the new `__init__` will need to read the log file for software determination, the folder or full path should be passed as well. In the current implementation the user can still just pass the file name.
If I misunderstood something, let's discuss it.

mjohnson541 replied:
Can't you just define self.path = os.path.join(os.getcwd(),path) in __init__?

alongd replied:
I don't think so, I have examples where the species.py file, which defines the Log file name, is in a different folder than Cantherm's input file, so `getcwd()` won't do the trick

alongd commented on rmgpy/cantherm/gaussian.py:
```diff
@@ -406,9 +406,8 @@ def loadNegativeFrequency(self):
         calculation in cm^-1.
         """
-        
+        frequency = None
```
We previously raised this error only for QChem, this commit adds it to Molpro and Gaussian (and slightly modifies it for QChem).
So yes, it changes when Cantherm will error

alongd commented on rmgpy/cantherm/kinetics.py:
```diff
@@ -266,21 +264,29 @@ def plot(self, outputDirectory):
-        for i in range(Tlist.shape[0]):
-            klist[i] = self.reaction.calculateTSTRateCoefficient(Tlist[i])
-            klist2[i] = self.reaction.kinetics.getRateCoefficient(Tlist[i])
+        t_list = [t for t in self.Tlist.value_si]
```
From the `__init__`:
```
        if Tlist is not None:
            self.Tlist = quantity.Quantity(Tlist)
```

alongd commented on rmgpy/cantherm/molpro.py:
```diff
@@ -195,6 +193,16 @@ def loadConformer(self, symmetry=None, spinMultiplicity=None, opticalIsomers=1,
         while line != '':
 
+            # Read the spin multiplicity if not explicitly given
+            if spinMultiplicity == 0 and 'spin' in line:
```
I couldn't find the spin multiplicity anywhere in a molpro log other that the input echo, e.g.:
```
 basis=6-311G(2d,d,p)
 int;
 wf,spin=0,charge=0;
 {uks,B3LYP}
```
which means that users who use the alternative format:
```
 basis=6-311G(2d,d,p)
 int;
 wf,33,1,1;
 {uks,B3LYP}
```
or anything else, would have to manually specify the multiplicity.
Let me know if you have a different recommendation for solving this.

cgrambow replied:
Ah, I see. Yeah, I usually don't specify spin in the input file because I mostly just have stable species or doublet radicals, which Molpro can detect automatically. Molpro prints `SPIN SYMMETRY: Singlet` somewhere in the output file for my jobs. We could parse that instead, assuming it's similar for doublets and triplets.

alongd commented on rmgpy/cantherm/thermo.py:
```diff
@@ -248,5 +247,10 @@ def plot(self, outputDirectory):
+        if not os.path.exists('plots'):
+            os.mkdir('plots')
+        valid_chars = "-_.()<=> %s%s" % (string.ascii_letters, string.digits)
+        filename = 'plots/' + ''.join(c for c in self.species.label if c in valid_chars) + '.pdf'
```
This line just manipulates a string. I do use `os.path.join` in the next line

cgrambow replied:
Yeah, you join `outputDirectory` and `filename`, but you should also use `os.path.join` to join `plots` and whatever follows instead of doing `'plots/' + ...`

alongd commented:
Thanks for your comments, @mjohnson541 and @cgrambow!
I added a bunch of fixup commits to address them (also rebased), please let me know whether you have additional comments, or if we're good to go.

alongd commented:
@mjohnson541, all examples run smooth now, thanks for pointing this out

alongd commented:
Rebased and squashed.
Could you re-approve?
